File: src/main/java/com/team195/CKSimMotor.java
package com.team195;

import edu.wpi.first.wpilibj.Sendable;
import edu.wpi.first.wpilibj.smartdashboard.SendableBuilder;
import edu.wpi.first.wpilibj.smartdashboard.SendableRegistry;

public class CKSimMotor implements Sendable {
    private int id;

    public CKSimMotor(int id) {
        if (!validateId(id)) {
            System.out.println("Invalid motor ID: " + id);
            return;
        }
        if (!CKSim.getInstance().registerMotor(id)) {
            System.out.println("Failed to add motor ID: " + id + "; ID Already in use!");
            return;
        }
        this.id = id;
        SendableRegistry.addLW(this, "CKSimMotor", id);
    }

    public float getMotorValue() {
        return CKSim.getInstance().getMotorValue(id);
    }

    public synchronized void setMotorValue(float value) {
        CKSim.getInstance().setMotorValue(id, value);
    }

    public float getEncoderValue() {
        return CKSim.getInstance().getEncoderValue(id);
    }

    private boolean validateId(int id) {
        return (id >= 0 && id < CKSim.MAX_NUM_MOTORS);
    }

    @Override
    public void initSendable(SendableBuilder builder) {
        builder.setSmartDashboardType("CKSimMotor");
        // builder.setActuator(false);
        // builder.setSafeState(this::setDisabled);
        builder.addDoubleProperty("Value", this::getMotorValue, value -> setMotorValue((float) value));
    }
}

File: src/main/native/cpp/CKSimMotor.cpp
#include "team195/CKSimMotor.h"
#include "frc/smartdashboard/SendableBuilder.h"
#include "frc/smartdashboard/SendableRegistry.h"

#include <iostream>

#include "team195/CKSim.h"

namespace team195
{
    CKSimMotor::CKSimMotor(int id) : id(id)
    {
        if (!validateId(id))
        {
            std::cout << "Invalid motor ID: " << id << std::endl;
            id = -1;
            return;
        }
        if (!CKSim::GetInstance().registerMotor(id))
        {
            std::cout << "Failed to add motor ID: " << id << "; ID Already in use!" << std::endl;
            id = -1;
            return;
        }
        frc::SendableRegistry::GetInstance().AddLW(this, "CKSimMotor", id);
    }

    float CKSimMotor::GetMotorValue()
    {
        return CKSim::GetInstance().getMotorValue(id);
    }

    void CKSimMotor::SetMotorValue(float value)
    {
        CKSim::GetInstance().setMotorValue(id, value);
    }

    float CKSimMotor::GetEncoderValue()
    {
        return CKSim::GetInstance().getEncoderValue(id);
    }

    bool CKSimMotor::validateId(int id)
    {
        return (id >= 0 && id < CKSim::MAX_NUM_MOTORS);
    }

    void CKSimMotor::InitSendable(frc::SendableBuilder &builder)
    {
        builder.SetSmartDashboardType("CKSimMotor");
        // builder.SetActuator(true);
        // builder.SetSafeState([=]() { SetDisabled(); });
        builder.AddDoubleProperty(
            "Value", [=]() { return (double)GetMotorValue(); },
            [=](double value) { SetMotorValue((float)value); });
    }
} // namespace team195